lib/Transforms/MemoryBanking.cpp

```diff
@@ -695,55 +765,168 @@ LogicalResult cleanUpOldMemRefs(DenseSet<Value> &oldMemRefVals,
+          funcOp.setArgAttr(argIndex, bankingDimensionsStr,
+                            getDimensionsAttr(specifiedOrDefaultDims));
+      }
+      return WalkResult::advance();
+    });
+  });
+}
+
 void MemoryBankingPass::runOnOperation() {
-  if (getOperation().isExternal() || bankingFactor == 1)
+  this->bankingFactors = {bankingFactorsList.begin(), bankingFactorsList.end()};
+  this->bankingDimensions = {bankingDimensionsList.begin(),
+                             bankingDimensionsList.end()};
+
+  if (getOperation().isExternal() ||
+      (bankingFactors.empty() ||
+       std::all_of(bankingFactors.begin(), bankingFactors.end(),
+                   [](unsigned f) { return f == 1; })))
     return;
 
-  if (bankingFactor == 0) {
+  if (std::any_of(bankingFactors.begin(), bankingFactors.end(),
+                  [](int f) { return f == 0; })) {
     getOperation().emitError("banking factor must be greater than 1");
     signalPassFailure();
     return;
   }
 
-  getOperation().walk([&](mlir::affine::AffineParallelOp parOp) {
-    DenseSet<Value> memrefsInPar = collectMemRefs(parOp);
-
-    for (auto memrefVal : memrefsInPar) {
-      auto [it, inserted] =
-          memoryToBanks.insert(std::make_pair(memrefVal, SmallVector<Value>{}));
-      if (inserted)
-        it->second = createBanks(memrefVal, bankingFactor, bankingDimension);
-    }
-  });
+  // `bankingFactors` is guaranteed to have elements and at least one of them is
+  // greater than 1 beyond this point.
+
+  setAllBankingAttributes(getOperation(), &getContext());
+
+  OpBuilder builder(getOperation());
+  while (true) {
+    memoryToBanks.clear();
+    oldMemRefVals.clear();
+    opsToErase.clear();
+
+    bool noBanksCreated = true;
+    getOperation().walk([&](mlir::affine::AffineParallelOp parOp) {
+      DenseSet<Value> memrefsInPar = collectMemRefs(parOp);
+      // We run `createBanks` iff there exists some `memrefVal` s.t. it has
+      // banking attributes attached to it.
+      for (auto memrefVal : memrefsInPar) {
+        auto currConfig = getMemRefBankingConfig(memrefVal);
+        if (!currConfig.factors) {
+          assert(!currConfig.dimensions &&
+                 "A banking dimension must be paired with a factor");
+          continue;
+        }
+        auto [it, inserted] = memoryToBanks.insert(
+            std::make_pair(memrefVal, SmallVector<Value>{}));
+        if (inserted)
+          it->second = createBanks(builder, memrefVal);
+        noBanksCreated = false;
+      }
+    });
+
+    if (failed(applyMemoryBanking(getOperation(), &getContext()))) {
+      signalPassFailure();
+    } else if (noBanksCreated)
+      // We stop the pass when IR stops changing:
+      // 1. there is no more banks are created;
+      // 2. `applyMemoryBanking` has reached its fixed point, which means every
+      // memory read/write operation has been rewritten to be using the newly
+      // created banks, and that the old memory references are erased.
```

Comment:
Nit: this seems a bit redundant with (1), I'm not totally in love with my change but it'd be nice to avoid saying the same thing twice.

```suggestion
      // We run this pass until convergence, i.e., `applyMemoryBanking` has reached its fixed point, which means every
      // memory read/write operation has been rewritten to be using the newly
      // created banks, and that the old memory references are erased.
```

Reply:
sure!